File: gatling-core/src/main/scala/io/gatling/core/scenario/Simulation.scala
/*
 * Copyright 2011-2021 GatlingCorp (https://gatling.io)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *  http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package io.gatling.core.scenario

import scala.concurrent.duration.FiniteDuration

import io.gatling.commons.stats.assertion.Assertion
import io.gatling.core.CoreComponents
import io.gatling.core.config.GatlingConfiguration
import io.gatling.core.controller.throttle.{ ThrottleStep, Throttling, Throttlings }
import io.gatling.core.pause._
import io.gatling.core.protocol.{ Protocol, ProtocolComponentsRegistries, Protocols }
import io.gatling.core.session.Expression
import io.gatling.core.structure.PopulationBuilder

abstract class Simulation {

  private var _populationBuilders: List[PopulationBuilder] = Nil
  private var _globalProtocols: Protocols = Map.empty
  private var _assertions = Seq.empty[Assertion]
  private var _maxDuration: Option[FiniteDuration] = None
  private var _globalPauseType: PauseType = Constant
  private var _globalThrottleSteps: Iterable[ThrottleStep] = Nil
  private var _beforeSteps: List[() => Unit] = Nil
  private var _afterSteps: List[() => Unit] = Nil

  def before(step: => Unit): Unit =
    _beforeSteps = _beforeSteps ::: List(() => step)

  def setUp(populationBuilders: PopulationBuilder*): SetUp = setUp(populationBuilders.toList)

  def setUp(populationBuilders: List[PopulationBuilder]): SetUp = {
    if (_populationBuilders.nonEmpty)
      throw new UnsupportedOperationException("setUp can only be called once")
    _populationBuilders = populationBuilders
    new SetUp
  }

  def after(step: => Unit): Unit =
    _afterSteps = _afterSteps ::: List(() => step)

  class SetUp {

    def protocols(ps: Protocol*): SetUp = protocols(ps.toIterable)

    def protocols(ps: Iterable[Protocol]): SetUp = {
      _globalProtocols = _globalProtocols ++ Protocol.indexByType(ps)
      this
    }

    def assertions(asserts: Assertion*): SetUp = assertions(asserts.toIterable)

    def assertions(asserts: Iterable[Assertion]): SetUp = {
      _assertions = _assertions ++ asserts
      this
    }

    def maxDuration(duration: FiniteDuration): SetUp = {
      _maxDuration = Some(duration)
      this
    }

    def throttle(throttleSteps: ThrottleStep*): SetUp = throttle(throttleSteps.toIterable)

    def throttle(throttleSteps: Iterable[ThrottleStep]): SetUp = {
      _globalThrottleSteps = throttleSteps
      this
    }

    def disablePauses: SetUp = pauses(Disabled)
    def constantPauses: SetUp = pauses(Constant)
    def exponentialPauses: SetUp = pauses(Exponential)
    def customPauses(custom: Expression[Long]): SetUp = pauses(new Custom(custom))
    def uniformPauses(plusOrMinus: Double): SetUp = pauses(new UniformPercentage(plusOrMinus))
    def uniformPauses(plusOrMinus: FiniteDuration): SetUp = pauses(new UniformDuration(plusOrMinus))
    def pauses(pauseType: PauseType): SetUp = {
      _globalPauseType = pauseType
      this
    }
  }

  private[gatling] def params(configuration: GatlingConfiguration): SimulationParams = {

    val rootPopulationBuilders = _populationBuilders
    require(rootPopulationBuilders.nonEmpty, "No scenario set up")

    val childrenPopulationBuilders = PopulationBuilder.groupChildrenByParent(rootPopulationBuilders)
    val allPopulationBuilders = rootPopulationBuilders ++ childrenPopulationBuilders.values.flatten

    val duplicates =
      allPopulationBuilders
        .groupBy(_.scenarioBuilder.name)
        .collect { case (name, scns) if scns.size > 1 => name }
    require(duplicates.isEmpty, s"Scenario names must be unique but found duplicates: $duplicates")

    allPopulationBuilders.foreach { scn =>
      require(scn.scenarioBuilder.name.nonEmpty, "Scenario name cannot be empty")
      require(scn.scenarioBuilder.actionBuilders.nonEmpty, s"Scenario ${scn.scenarioBuilder.name} is empty")
    }

    val scenarioThrottlings: Map[String, Throttling] = allPopulationBuilders.flatMap { populationBuilder =>
      val steps = populationBuilder.scenarioThrottleSteps

      if (steps.isEmpty) {
        Nil
      } else {
        List(populationBuilder.scenarioBuilder.name -> Throttling(steps))
      }
    }.toMap

    val globalThrottling =
      if (_globalThrottleSteps.isEmpty) {
        None
      } else {
        Some(Throttling(_globalThrottleSteps))
      }

    val maxDuration = {
      val globalThrottlingMaxDuration = globalThrottling.map(_.duration)
      val scenarioThrottlingMaxDurations = scenarioThrottlings.values.map(_.duration).toList

      _maxDuration.map(List(_)).getOrElse(Nil) ::: globalThrottlingMaxDuration.map(List(_)).getOrElse(Nil) ::: scenarioThrottlingMaxDurations match {
        case Nil => None
        case nel => Some(nel.min)
      }
    }

    new SimulationParams(
      getClass.getName,
      rootPopulationBuilders,
      childrenPopulationBuilders,
      _globalProtocols,
      _globalPauseType,
      Throttlings(globalThrottling, scenarioThrottlings),
      maxDuration,
      _assertions
    )
  }

  private[gatling] def executeBefore(): Unit = _beforeSteps.foreach(_.apply())
  private[gatling] def executeAfter(): Unit = _afterSteps.foreach(_.apply())
}

final class SimulationParams(
    val name: String,
    val rootPopulationBuilders: List[PopulationBuilder],
    val childrenPopulationBuilders: Map[String, List[PopulationBuilder]],
    val globalProtocols: Protocols,
    val globalPauseType: PauseType,
    val throttlings: Throttlings,
    val maxDuration: Option[FiniteDuration],
    val assertions: Seq[Assertion]
) {

  private def buildScenario(populationBuilder: PopulationBuilder, coreComponents: CoreComponents, protocolComponentsRegistries: ProtocolComponentsRegistries) =
    populationBuilder.build(coreComponents, protocolComponentsRegistries, globalPauseType, throttlings.global)

  def scenarios(coreComponents: CoreComponents): Scenarios = {
    val protocolComponentsRegistries = new ProtocolComponentsRegistries(coreComponents, globalProtocols)
    val rootScenarios = rootPopulationBuilders.map(buildScenario(_, coreComponents, protocolComponentsRegistries))
    val childrenScenarios = childrenPopulationBuilders.view.mapValues(_.map(buildScenario(_, coreComponents, protocolComponentsRegistries))).toMap

    new Scenarios(rootScenarios, childrenScenarios)
  }
}

File: gatling-core-java/src/main/java/io/gatling/core/javaapi/Simulation.java
/*
 * Copyright 2011-2021 GatlingCorp (https://gatling.io)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *  http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package io.gatling.core.javaapi;

import io.gatling.core.config.GatlingConfiguration;
import io.gatling.core.protocol.Protocol;
import io.gatling.core.scenario.Simulation$;
import io.gatling.core.scenario.SimulationParams;
import scala.Option;

import java.time.Duration;
import java.util.*;
import java.util.function.Function;
import java.util.stream.Collectors;

import static io.gatling.core.javaapi.internal.ScalaHelpers.*;

public abstract class Simulation {

  private List<PopulationBuilder> _populationBuilders;
  private Map<Class<? extends Protocol>, Protocol> _globalProtocols = new HashMap<>();
  private List<Assertion> _assertions = new ArrayList<>();
  private Duration _maxDuration;
  private PauseType _globalPauseType = PauseType.Constant;
  private List<ThrottleStep> _globalThrottleSteps = new ArrayList<>();

  public void before() {
  }

  public void after() {
  }


  public SetUp setUp(PopulationBuilder... populationBuilders) {
    return setUp(Arrays.asList(populationBuilders));
  }

  public SetUp setUp(List<PopulationBuilder> populationBuilders) {
    if (_populationBuilders != null) {
      throw new UnsupportedOperationException("Can only call setUp once");
    }
    _populationBuilders = populationBuilders;
    return new SetUp();
  }

  public class SetUp {

    private SetUp() {
    }

    public SetUp protocols(ProtocolBuilder... protocols) {
      return protocols(Arrays.asList(protocols));
    }

    public SetUp protocols(List<ProtocolBuilder> protocols) {
      _globalProtocols = protocols.stream().map(ProtocolBuilder::protocol).collect(Collectors.toMap(Protocol::getClass, Function.identity()));
      return this;
    }

    public SetUp assertions(Assertion... assertions) {
      return assertions(Arrays.asList(assertions));
    }

    public SetUp assertions(List<Assertion> assertions) {
      _assertions = assertions;
      return this;
    }

    public SetUp maxDuration(int duration) {
      return maxDuration(Duration.ofSeconds(duration));
    }

    public SetUp maxDuration(Duration duration) {
      _maxDuration = duration;
      return this;
    }

    public SetUp throttle(ThrottleStep... throttleSteps) {
      return throttle(Arrays.asList(throttleSteps));
    }

    public SetUp throttle(List<ThrottleStep> throttleSteps) {
      _globalThrottleSteps = throttleSteps;
      return this;
    }

    public SetUp disablePauses() {
      return pauses(PauseType.Disabled);
    }

    public SetUp constantPauses() {
      return pauses(PauseType.Constant);
    }

    public SetUp exponentialPauses() {
      return pauses(PauseType.Exponential);
    }

    public SetUp customPauses(Function<Session, Long> f) {
      return pauses(new PauseType.Custom(toUntypedGatlingSessionFunction(f)));
    }

    public SetUp uniformPauses(double plusOrMinus) {
      return pauses(new PauseType.UniformPercentage(plusOrMinus));
    }

    public SetUp uniformPauses(Duration plusOrMinus) {
      return pauses(new PauseType.UniformDuration(plusOrMinus));
    }

    public SetUp pauses(PauseType pauseType) {
      _globalPauseType = pauseType;
      return this;
    }
  }

  public SimulationParams params(GatlingConfiguration configuration) {
    return Simulation$.MODULE$.params(
      getClass().getName(),
      toScalaSeq(_populationBuilders.stream().map(PopulationBuilder::asScala).collect(Collectors.toList())).toList(),
      toScalaMap(_globalProtocols),
      toScalaSeq(_assertions.stream().map(Assertion::asScala).collect(Collectors.toList())),
      Option.apply(_maxDuration).map(duration -> toScalaDuration(duration)),
      _globalPauseType.asScala(),
      toScalaSeq(_globalThrottleSteps.stream().map(ThrottleStep::asScala).collect(Collectors.toList())),
      () -> {
        before();
        return null;
      },
      () -> {
        after();
        return null;
      },
      configuration
    );
  }
}
